Add support for local field validation for synced bookmarks and credentials

ayoy opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations.
-->

Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/0/1206637913741251/f
iOS PR: https://github.com/duckduckgo/iOS/pull/2614
macOS PR: https://github.com/duckduckgo/macos-browser/pull/2454
What kind of version bump will this require?: Major

**Description**:
This change adds a mechanism that filters out syncable objects that would fail validation on the backend
before sending Sync patch request. Objects rejected from patch payload are retried on every subsequent
Sync request, until they're updated to pass validation or deleted.

<!--
Tagging instructions
If this PR isn't ready to be merged for whatever reason it should be marked with the `DO NOT MERGE` label (particularly if it's a draft)
If it's pending Product Review/PFR, please add the `Pending Product Review` label.

If at any point it isn't actively being worked on/ready for review/otherwise moving forward (besides the above PR/PFR exception) strongly consider closing it (or not opening it in the first place). If you decide not to close it, make sure it's labelled to make it clear the PRs state and comment with more information.
-->

**Steps to test this PR**:
1. Go through testing scenarios in [this Asana task](https://app.asana.com/0/72649045549333/1206754264327477/f). Have a look at [this comment](https://app.asana.com/0/0/1206754264327477/1206880070407408/f) for an exception for Scenario B.
2. Smoke test Sync overall
3. Go to `DDGSync.swift` and set `isFieldValidationEnabled` to `false`; verify that field validation is disabled and Sync works correctly without field validation. Feel free to commit the change to remote to simplify testing client apps.

<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable.
-->

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

bwaresiak commented on Sources/Bookmarks/BookmarkUtils.swift:
```diff
@@ -174,6 +174,16 @@ public struct BookmarkUtils {
     }
 
+    public static func fetchTitlesForBookmarks(modifiedBefore date: Date, in context: NSManagedObjectContext) -> [String] {
+        let request = NSFetchRequest<NSFetchRequestResult>(entityName: "BookmarkEntity")
```
You can use `BookmarkEntity.fetchRequest()` to avoid hardcoded strings.

ayoy replied:
Yep, I saw it but that one is for `BookmarkEntity` result type, and for dictionary result type we need `NSFetchRequestResult`.

bwaresiak commented on Sources/SyncDataProviders/Bookmarks/BookmarksProvider.swift:
```diff
@@ -82,7 +98,22 @@ public final class BookmarksProvider: DataProvider {
+                        os_log(
+                            .error,
+                            log: log,
+                            "Validation failed for bookmark %{public}s with title: %{public}s",
```
Since title is user data, I believe it should be `%{private}`

ayoy replied:
For some reason I thought `%{private}` wouldn't show up in Xcode console (while it's only true about Console.app) and I set it for debugging 🤦

bwaresiak commented on Sources/SyncDataProviders/Bookmarks/internal/SyncableBookmarkAdapter.swift:
```diff
@@ -84,10 +92,15 @@ extension Syncable {
+                    } else {
+                        payload["title"] = try encrypt(title)
+                    }
+                }
```
I'm tempted to just release this as-is (ignoring the flag), as it doesn't really affect object graph.

ayoy replied:
You specifically mean truncating folder titles, right? Just this?

bwaresiak replied:
Yes, just truncating.

ayoy replied:
Sure, it's done now.

bwaresiak commented on Sources/SyncDataProviders/Credentials/CredentialsProvider.swift:
```diff
@@ -78,15 +79,35 @@ public final class CredentialsProvider: DataProvider {
+                os_log(
+                    .error,
+                    log: log,
+                    "Validation failed for credential %{public}s with title: %{public}s",
```
Same here, `%{private}`

bwaresiak reviewed (commented):
Great job, few comments inline.

ayoy commented:
@bwaresiak I've addressed your comments and added the mechanism to not mark a bookmark as modified when only its parent or favorites folders change. It's ready for you to have another look. Thanks!

bwaresiak reviewed (approved):
Great job!
